Enforce -Wunsafe-buffer-usage in Source/WebGPU

geoffreygaren opened the pull request:
#### 5138438b500796279ab122bbcce0c1af840579d0
<pre>
Enforce -Wunsafe-buffer-usage in Source/WebGPU
<a href="https://bugs.webkit.org/show_bug.cgi?id=281792">https://bugs.webkit.org/show_bug.cgi?id=281792</a>
<a href="https://rdar.apple.com/138218909">rdar://138218909</a>

Reviewed by Mike Wyrzykowski.

Added an IndexedRange class so you can access the iteration index in
a safe range-based for loop:

    for (auto [ index, value ] : IndexedRange(collection)) { ... }

Deployed std::array and std::span in lots of places.

Added a parseHexDouble helper because, hilariously, it is the year 2024
and there is no bounds-check aware option for hex parsing of 16bit
characters in C or C++.

I chose to make xxxSpan() member functions for most WGPU API types, rather than
modifying the types themselves, in order to keep the existing API stable
and reduce the scope of change.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/IndexedRange.h: Added.
(WTF::IndexedRangeIterator::IndexedRangeIterator):
(WTF::IndexedRangeIterator::operator++):
(WTF::IndexedRangeIterator::operator*):
(WTF::IndexedRangeIterator::operator== const):
(WTF::IndexedRange::IndexedRange):
(WTF::IndexedRange::begin):
(WTF::IndexedRange::end):
* Source/WTF/wtf/ReferenceWrapperVector.h:
(WTF::ReferenceWrapperVectorIterator::operator++):
(WTF::ReferenceWrapperVectorIterator::operator--):
(WTF::ReferenceWrapperVectorConstIterator::operator++):
(WTF::ReferenceWrapperVectorConstIterator::operator--):
* Source/WTF/wtf/dtoa/utils.h:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/WGSL/AST/ASTBinaryExpression.h:
(WGSL::AST::toASCIILiteral):
* Source/WebGPU/WGSL/AST/ASTBuilder.cpp:
(WGSL::AST::Builder::Builder):
(WGSL::AST::Builder::allocateArena):
(WGSL::AST::Builder::saveCurrentState):
(WGSL::AST::Builder::restore):
(WGSL::AST::Builder::arena): Deleted.
* Source/WebGPU/WGSL/AST/ASTBuilder.h:
(WGSL::AST::Builder::construct):
* Source/WebGPU/WGSL/AST/ASTUnaryExpression.h:
(WGSL::AST::toASCIILiteral):
* Source/WebGPU/WGSL/MangleNames.cpp:
(WGSL::MangledName::toString const):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::maybeSplitToken):
* Source/WebGPU/WGSL/Types.h:
(WGSL::Types::PrimitiveStruct::&gt;&gt;):
* Source/WebGPU/WebGPU/Adapter.mm:
(WebGPU::Adapter::requestDevice):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createExternalTextureFromPixelBuffer const):
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindGroupLayout.h:
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::errorValidatingDynamicOffsets const):
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::span):
(WebGPU::Buffer::getMappedRange):
(WebGPU::Buffer::getBufferContents):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::setBindGroup):
(wgpuComputePassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(WebGPU::Device::createPipelineLayout):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::configure):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
(wgpuQueueSubmit):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::updateMinMaxDepths):
(WebGPU::RenderBundle::validateRenderPass const):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::makeRenderBundleICBWithResources):
(WebGPU::Device::createRenderBundleEncoder):
(WebGPU::makeSpanFromBuffer):
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::m_maxDrawCount):
(WebGPU::RenderPassEncoder::setBindGroup):
(wgpuRenderPassEncoderExecuteBundles):
(wgpuRenderPassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createVertexDescriptor):
(WebGPU::Device::addPipelineLayouts):
(WebGPU::RenderPipeline::validateRenderBundle const):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
* Source/WebGPU/WebGPU/WebGPU.h:
(span):

Canonical link: <a href="https://commits.webkit.org/285506@main">https://commits.webkit.org/285506@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/442c421938d810c61911c14da7115ef8cac66622

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72832 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77028 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24066 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60062 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23882 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57262 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15756 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75899 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47243 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62683 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37691 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43890 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20150 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22395 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65965 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65743 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20507 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78701 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72088 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17078 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19650 "Found 2 new test failures: http/wpt/mediarecorder/pause-recording.html imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65711 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62690 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64987 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16064 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13300 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6952 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93867 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48055 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20652 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49122 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50417 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48867 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -864,7 +863,7 @@ static auto makeBindingLayout(WGPUBindGroupLayoutEntry& newEntry, auto& bindingM
             }); existingBindingIndex != notFound) {
                 entries[existingBindingIndex].visibility |= visibility;
-                entries[existingBindingIndex].metalBinding[stage] = entry.binding;
+                std::span(entries[existingBindingIndex].metalBinding)[stage] = entry.binding;
```
could also change the definition in WebGPU.h from:
```
uint32_t metalBinding[WGPUShaderStage_Compute / 2 + 1];
```
to
```
std::array<uint32_t, 3> metalBinding;
```
to avoid adding the `std::span` here

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -1767,7 +1764,7 @@ static uint32_t componentsForDataType(MTLDataType dataType)
     auto& fragment = *m_descriptor.fragment;
     for (size_t i = 0, maxTargetCount = std::max<size_t>(fragment.targetCount, descriptor.colorFormatCount); i < maxTargetCount; ++i) {
-        auto colorFormat = i < descriptor.colorFormatCount ? descriptor.colorFormats[i] : WGPUTextureFormat_Undefined;
+        auto colorFormat = i < descriptor.colorFormatCount ? span(descriptor)[i] : WGPUTextureFormat_Undefined;
```
It seems a little odd that `span(descriptor)[i]` returns the color format, since descriptor may have two or more arrays

mwyrzykowski commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1866,8 +1866,27 @@ WGPU_EXPORT WGPUBool wgpuXRViewIsValid(WGPUXRView view) WGPU_FUNCTION_ATTRIBUTE;
+inline auto span(const WGPUBindGroupDescriptor& bindGroupDescriptor) { return unsafeForgeSpan(bindGroupDescriptor.entries, bindGroupDescriptor.entryCount); }
+inline auto span(const WGPUBindGroupLayoutDescriptor& bindGroupLayoutDescriptor) { return unsafeForgeSpan(bindGroupLayoutDescriptor.entries, bindGroupLayoutDescriptor.entryCount); }
+inline auto span(const WGPURenderPassDescriptor& renderPassDescriptor) { return unsafeForgeSpan(renderPassDescriptor.colorAttachments, renderPassDescriptor.colorAttachmentCount); }
+inline auto span(const WGPURenderBundleEncoderDescriptor& renderBundleEncoderDescriptor) { return unsafeForgeSpan(renderBundleEncoderDescriptor.colorFormats, renderBundleEncoderDescriptor.colorFormatCount); }
```
Would propose naming this `colorFormatsSpan` since it is not obvious from the name what container is being returned as a span.

Similar suggestions apply for the other `span` functions where it is not obvious from the function parameter which container is being used to create the span

mwyrzykowski reviewed (approved):
This looks great - thank you!

Confirmed the test failures are unrelated

geoffreygaren committed:
Enforce -Wunsafe-buffer-usage in Source/WebGPU
https://bugs.webkit.org/show_bug.cgi?id=281792
rdar://138218909

Reviewed by Mike Wyrzykowski.

Added an IndexedRange class so you can access the iteration index in
a safe range-based for loop:

    for (auto [ index, value ] : IndexedRange(collection)) { ... }

Deployed std::array and std::span in lots of places.

Added a parseHexDouble helper because, hilariously, it is the year 2024
and there is no bounds-check aware option for hex parsing of 16bit
characters in C or C++.

I chose to make xxxSpan() member functions for most WGPU API types, rather than
modifying the types themselves, in order to keep the existing API stable
and reduce the scope of change.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/IndexedRange.h: Added.
(WTF::IndexedRangeIterator::IndexedRangeIterator):
(WTF::IndexedRangeIterator::operator++):
(WTF::IndexedRangeIterator::operator*):
(WTF::IndexedRangeIterator::operator== const):
(WTF::IndexedRange::IndexedRange):
(WTF::IndexedRange::begin):
(WTF::IndexedRange::end):
* Source/WTF/wtf/ReferenceWrapperVector.h:
(WTF::ReferenceWrapperVectorIterator::operator++):
(WTF::ReferenceWrapperVectorIterator::operator--):
(WTF::ReferenceWrapperVectorConstIterator::operator++):
(WTF::ReferenceWrapperVectorConstIterator::operator--):
* Source/WTF/wtf/dtoa/utils.h:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/WGSL/AST/ASTBinaryExpression.h:
(WGSL::AST::toASCIILiteral):
* Source/WebGPU/WGSL/AST/ASTBuilder.cpp:
(WGSL::AST::Builder::Builder):
(WGSL::AST::Builder::allocateArena):
(WGSL::AST::Builder::saveCurrentState):
(WGSL::AST::Builder::restore):
(WGSL::AST::Builder::arena): Deleted.
* Source/WebGPU/WGSL/AST/ASTBuilder.h:
(WGSL::AST::Builder::construct):
* Source/WebGPU/WGSL/AST/ASTUnaryExpression.h:
(WGSL::AST::toASCIILiteral):
* Source/WebGPU/WGSL/MangleNames.cpp:
(WGSL::MangledName::toString const):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::maybeSplitToken):
* Source/WebGPU/WGSL/Types.h:
(WGSL::Types::PrimitiveStruct::>>):
* Source/WebGPU/WebGPU/Adapter.mm:
(WebGPU::Adapter::requestDevice):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createExternalTextureFromPixelBuffer const):
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindGroupLayout.h:
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::errorValidatingDynamicOffsets const):
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::span):
(WebGPU::Buffer::getMappedRange):
(WebGPU::Buffer::getBufferContents):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::setBindGroup):
(wgpuComputePassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(WebGPU::Device::createPipelineLayout):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::configure):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::writeTexture):
(wgpuQueueSubmit):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(WebGPU::RenderBundle::updateMinMaxDepths):
(WebGPU::RenderBundle::validateRenderPass const):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::makeRenderBundleICBWithResources):
(WebGPU::Device::createRenderBundleEncoder):
(WebGPU::makeSpanFromBuffer):
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::m_maxDrawCount):
(WebGPU::RenderPassEncoder::setBindGroup):
(wgpuRenderPassEncoderExecuteBundles):
(wgpuRenderPassEncoderSetBindGroup):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createVertexDescriptor):
(WebGPU::Device::addPipelineLayouts):
(WebGPU::RenderPipeline::validateRenderBundle const):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
* Source/WebGPU/WebGPU/WebGPU.h:
(span):

Canonical link: https://commits.webkit.org/285506@main
